contribute/design/proposal-cstor-orchestration.md

```diff
@@ -0,0 +1,57 @@
+# Managing cStor based OpenEBS Volumes. 
+
+## Background 
+
+
+## Design
+
+Provisioning of cStor based Volume involve two steps. 
+
+1. Creating cStor StoragePools:
+
+   * Admin will create SPC that contains information like:
+     - type=block
+     - name
+     - disks list (type of PDC or specific disks)
+     - resource requests (optional, like CPU, RAM)
+     - raid type (optional, default _stripe_)
+     - nodes list (optional, default: _all_)
+     - max-pools (optional, default: _none_)
+
+   * maya-cstor-operator (could be embedded into maya-apiserver), will be watching for SPCs (type=block).
+   
+   * When maya-cstor-operator detects a new SPC object, it will identify the list of nodes that satisfy the SPC constraints in terms of:
+     - availability of disks
+     - resources (CPU and RAM) 
+     - nodes filter. 
+     This step can result in more than one node satisfying the constraints. Only the number of nodes required (as specified using maz-pools) will be picked up. 
+     
+   * For each of the potentional node where the cStor pool can be created, maya-cstor-operator will:
+     * create a cStorPool (CRD), which will include the following information:
+       - unique id
+       - name
+       - actual disks paths to be used. 
+       - raid type
+     * create a Deployment YAML file that contains the cStor container and its associated sidecars. The cStor-Sidecar is passed the “unique id” of the cStorPool (CRD) 
+     * create SP object
+
+
+2. Creating Volume using the (cStor) Storage Pools:
+
+   * Admin will create a PVC that is associated to StoragePool (linked by the pool name). 
+
+   * maya-apiserver will create the OV Controller and the OV Service (to get the portal ip address). The OV Controller will be passed a ConfigMap (that will be mounted into the OV Controller). The ConfigMap will contain the following details:
+     - portal ip
+     - ? 
+
+   * maya-apiserver will create the OV Replicas as follows: 
+     * Query for the storage pools matching the given pool name in the PVC and pick up a subset of pools (based on the replica-count of the PVC). 
+     * For each replica, maya-apiserver will create - cStorVolume CRD. This cStorVolume CRD will contain:
+       - cStorPool (unique id)
+       - Unique Name
+       - Required capacity
+       - OV Controller IP
+       
+    * The cstor-sidecar running in the cStorPool(unique id), will act on the cStorVolume CRD for creating the zVol and associating itself with the OV Controller.
```

Comment:
can we mention `watch` instead of `act`?

side-car will be responsible for update to the cStorVolume CR. It should NOT be able to create or delete.

Reply:
Will cstor container be able to give immediate responses for all kinds of actions?

Reply:
yes sir.. we can move ahead with this assumption.